File: infra-market-server-go/internal/service/activity_component_service.go
package service

import (
	"encoding/json"
	"log"
	"net/http"

	"github.com/bucketheadv/infra-market/internal/dto"
	"github.com/bucketheadv/infra-market/internal/entity"
	"github.com/bucketheadv/infra-market/internal/repository"
	"github.com/bucketheadv/infra-market/internal/util"
	"gorm.io/gorm"
)

type ActivityComponentService struct {
	db              *gorm.DB
	componentRepo   *repository.ActivityComponentRepository
}

func NewActivityComponentService(
	db *gorm.DB,
	componentRepo *repository.ActivityComponentRepository,
) *ActivityComponentService {
	return &ActivityComponentService{
		db:              db,
		componentRepo:   componentRepo,
	}
}

// GetActivityComponents 获取活动组件列表
func (s *ActivityComponentService) GetActivityComponents(query dto.ActivityComponentQueryDto) dto.ApiData[dto.PageResult[dto.ActivityComponentDto]] {
	components, total, err := s.componentRepo.Page(query)
	if err != nil {
		return dto.Error[dto.PageResult[dto.ActivityComponentDto]]("查询失败", http.StatusInternalServerError)
	}

	// 转换为DTO
	componentDtos := make([]dto.ActivityComponentDto, len(components))
	for i, component := range components {
		componentDtos[i] = s.convertComponentToDto(&component)
	}

	result := dto.PageResult[dto.ActivityComponentDto]{
		Records: componentDtos,
		Total:   total,
		Page:    query.GetPage(),
		Size:    query.GetSize(),
	}

	return dto.Success(result)
}

// GetAllActivityComponents 获取所有活动组件
func (s *ActivityComponentService) GetAllActivityComponents() dto.ApiData[[]dto.ActivityComponentDto] {
	components, err := s.componentRepo.List()
	if err != nil {
		return dto.Error[[]dto.ActivityComponentDto]("查询失败", http.StatusInternalServerError)
	}

	// 转换为DTO
	componentDtos := make([]dto.ActivityComponentDto, len(components))
	for i, component := range components {
		componentDtos[i] = s.convertComponentToDto(&component)
	}

	return dto.Success(componentDtos)
}

// GetActivityComponent 获取活动组件详情
func (s *ActivityComponentService) GetActivityComponent(id uint64) dto.ApiData[dto.ActivityComponentDto] {
	component, err := s.componentRepo.FindByID(id)
	if err != nil {
		log.Printf("获取活动组件详情失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityComponentDto]("活动组件不存在", http.StatusNotFound)
	}

	componentDto := s.convertComponentToDto(component)
	return dto.Success(componentDto)
}

// CreateActivityComponent 创建活动组件
func (s *ActivityComponentService) CreateActivityComponent(form dto.ActivityComponentFormDto) dto.ApiData[dto.ActivityComponentDto] {
	// 序列化字段/组件配置
	fieldsJSON, err := s.serializeFields(form.Fields)
	if err != nil {
		log.Printf("序列化字段/组件配置失败: %v\n", err)
		return dto.Error[dto.ActivityComponentDto]("字段/组件配置格式错误", http.StatusBadRequest)
	}

	status := 1
	if form.Status != nil {
		status = *form.Status
	}

	component := &entity.ActivityComponent{
		Name:        form.Name,
		Description: form.Description,
		Fields:      fieldsJSON,
		Status:      status,
	}

	if err := s.componentRepo.Create(component); err != nil {
		log.Printf("创建活动组件失败: %v\n", err)
		return dto.Error[dto.ActivityComponentDto]("创建活动组件失败", http.StatusInternalServerError)
	}

	componentDto := s.convertComponentToDto(component)
	return dto.Success(componentDto)
}

// UpdateActivityComponent 更新活动组件
func (s *ActivityComponentService) UpdateActivityComponent(id uint64, form dto.ActivityComponentFormDto) dto.ApiData[dto.ActivityComponentDto] {
	component, err := s.componentRepo.FindByID(id)
	if err != nil {
		log.Printf("更新活动组件失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityComponentDto]("活动组件不存在", http.StatusNotFound)
	}

	// 序列化字段/组件配置
	var fieldsJSON *string
	if form.Fields != nil {
		jsonStr, err := s.serializeFields(form.Fields)
		if err != nil {
			log.Printf("序列化字段/组件配置失败: %v\n", err)
			return dto.Error[dto.ActivityComponentDto]("字段/组件配置格式错误", http.StatusBadRequest)
		}
		fieldsJSON = jsonStr
	}

	component.Name = form.Name
	if form.Description != nil {
		component.Description = form.Description
	}
	if fieldsJSON != nil {
		component.Fields = fieldsJSON
	}
	if form.Status != nil {
		component.Status = *form.Status
	}

	if err := s.componentRepo.Update(component); err != nil {
		log.Printf("更新活动组件失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityComponentDto]("更新活动组件失败", http.StatusInternalServerError)
	}

	componentDto := s.convertComponentToDto(component)
	return dto.Success(componentDto)
}

// DeleteActivityComponent 删除活动组件
func (s *ActivityComponentService) DeleteActivityComponent(id uint64) dto.ApiData[any] {
	_, err := s.componentRepo.FindByID(id)
	if err != nil {
		log.Printf("删除活动组件失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[any]("活动组件不存在", http.StatusNotFound)
	}

	if err := s.componentRepo.Delete(id); err != nil {
		log.Printf("删除活动组件失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[any]("删除活动组件失败", http.StatusInternalServerError)
	}

	return dto.Success[any](nil)
}

// UpdateActivityComponentStatus 更新活动组件状态
func (s *ActivityComponentService) UpdateActivityComponentStatus(id uint64, status int) dto.ApiData[dto.ActivityComponentDto] {
	component, err := s.componentRepo.FindByID(id)
	if err != nil {
		log.Printf("更新活动组件状态失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityComponentDto]("活动组件不存在", http.StatusNotFound)
	}

	component.Status = status
	if err := s.componentRepo.Update(component); err != nil {
		log.Printf("更新活动组件状态失败，组件ID: %d, 状态: %d, 错误: %v\n", id, status, err)
		return dto.Error[dto.ActivityComponentDto]("更新状态失败", http.StatusInternalServerError)
	}

	componentDto := s.convertComponentToDto(component)
	return dto.Success(componentDto)
}

// CopyActivityComponent 复制活动组件
func (s *ActivityComponentService) CopyActivityComponent(id uint64) dto.ApiData[dto.ActivityComponentDto] {
	component, err := s.componentRepo.FindByID(id)
	if err != nil {
		log.Printf("复制活动组件失败，组件ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityComponentDto]("活动组件不存在", http.StatusNotFound)
	}

	newComponent := &entity.ActivityComponent{
		Name:        component.Name + "_副本",
		Description: component.Description,
		Fields:      component.Fields,
		Status:      0,
	}

	if err := s.componentRepo.Create(newComponent); err != nil {
		log.Printf("复制活动组件失败: %v\n", err)
		return dto.Error[dto.ActivityComponentDto]("复制活动组件失败", http.StatusInternalServerError)
	}

	componentDto := s.convertComponentToDto(newComponent)
	return dto.Success(componentDto)
}

// convertComponentToDto 转换活动组件实体为DTO
func (s *ActivityComponentService) convertComponentToDto(component *entity.ActivityComponent) dto.ActivityComponentDto {
	fields := s.parseFields(component.Fields)

	return dto.ActivityComponentDto{
		ID:          component.ID,
		Name:        component.Name,
		Description: component.Description,
		Fields:      fields,
		Status:      component.Status,
		CreateTime:  util.Format(&component.CreateTime),
		UpdateTime:  util.Format(&component.UpdateTime),
	}
}

// serializeFields 序列化字段/组件配置为JSON字符串
func (s *ActivityComponentService) serializeFields(fields []dto.ActivityComponentFieldDto) (*string, error) {
	if fields == nil {
		return nil, nil
	}

	jsonBytes, err := json.Marshal(fields)
	if err != nil {
		return nil, err
	}

	jsonStr := string(jsonBytes)
	return &jsonStr, nil
}

// parseFields 解析字段/组件配置JSON字符串
func (s *ActivityComponentService) parseFields(fieldsJSON *string) []dto.ActivityComponentFieldDto {
	if fieldsJSON == nil || *fieldsJSON == "" {
		return nil
	}

	var fields []dto.ActivityComponentFieldDto
	if err := json.Unmarshal([]byte(*fieldsJSON), &fields); err != nil {
		log.Printf("解析字段/组件配置失败: %v\n", err)
		return nil
	}

	return fields
}

File: infra-market-server-vertx/src/main/kotlin/io/infra/market/vertx/service/ActivityComponentService.kt
package io.infra.market.vertx.service

import com.google.inject.Inject
import com.google.inject.Singleton
import io.infra.market.vertx.config.JacksonConfig
import io.infra.market.vertx.dto.ActivityComponentDto
import io.infra.market.vertx.dto.ActivityComponentFormDto
import io.infra.market.vertx.dto.ActivityComponentQueryDto
import io.infra.market.vertx.dto.ApiData
import io.infra.market.vertx.dto.PageResultDto
import io.infra.market.vertx.entity.ActivityComponent
import io.infra.market.vertx.repository.ActivityComponentDao
import io.infra.market.vertx.util.TimeUtil
import org.slf4j.LoggerFactory

/**
 * 活动组件服务
 * 
 * 规则1：任何调用 xxx.awaitForResult() 的函数，必须用 suspend 修饰
 */
@Singleton
class ActivityComponentService @Inject constructor(
    private val activityComponentDao: ActivityComponentDao
) {
    
    private val logger = LoggerFactory.getLogger(ActivityComponentService::class.java)
    
    suspend fun getActivityComponents(query: ActivityComponentQueryDto): ApiData<PageResultDto<ActivityComponentDto>> {
        val page = query.page ?: 1
        val size = query.size ?: 10
        
        val pageResult = activityComponentDao.findPage(
            name = query.name,
            status = query.status,
            page = page,
            size = size
        )
        
        val componentDtos = pageResult.records.map { convertToDto(it) }
        
        val result = PageResultDto(
            records = componentDtos,
            total = pageResult.total,
            page = pageResult.page,
            size = pageResult.size
        )
        
        return ApiData.success(result)
    }
    
    suspend fun getAllActivityComponents(): ApiData<List<ActivityComponentDto>> {
        val components = activityComponentDao.findAll()
        val componentDtos = components.map { convertToDto(it) }
        // 只返回启用的组件
        val enabledComponents = componentDtos.filter { it.status == 1 }
        return ApiData.success(enabledComponents)
    }
    
    suspend fun getActivityComponent(id: Long): ApiData<ActivityComponentDto> {
        val component = activityComponentDao.findById(id) ?: return ApiData.error("活动组件不存在", 404)
        
        val componentDto = convertToDto(component)
        return ApiData.success(componentDto)
    }
    
    suspend fun createActivityComponent(form: ActivityComponentFormDto): ApiData<ActivityComponentDto> {
        val component = convertToEntity(form)
        component.status = form.status ?: 1
        
        val componentId = activityComponentDao.save(component)
        component.id = componentId
        
        val componentDto = convertToDto(component)
        return ApiData.success(componentDto)
    }
    
    suspend fun updateActivityComponent(id: Long, form: ActivityComponentFormDto): ApiData<ActivityComponentDto> {
        val existingComponent = activityComponentDao.findById(id) ?: return ApiData.error("活动组件不存在", 404)
        
        existingComponent.name = form.name
        existingComponent.description = form.description
        existingComponent.fields = serializeFields(form.fields)
        if (form.status != null) {
            existingComponent.status = form.status
        }
        
        activityComponentDao.updateById(existingComponent)
        
        val componentDto = convertToDto(existingComponent)
        return ApiData.success(componentDto)
    }
    
    suspend fun deleteActivityComponent(id: Long): ApiData<Unit> {
        activityComponentDao.findById(id) ?: return ApiData.error("活动组件不存在", 404)
        
        activityComponentDao.deleteById(id)
        
        return ApiData.success()
    }
    
    suspend fun updateActivityComponentStatus(id: Long, status: Int): ApiData<ActivityComponentDto> {
        val component = activityComponentDao.findById(id) ?: return ApiData.error("活动组件不存在", 404)
        
        component.status = status
        activityComponentDao.updateById(component)
        
        val componentDto = convertToDto(component)
        return ApiData.success(componentDto)
    }
    
    suspend fun copyActivityComponent(id: Long): ApiData<ActivityComponentDto> {
        val existingComponent = activityComponentDao.findById(id) ?: return ApiData.error("活动组件不存在", 404)
        
        val newComponent = existingComponent.copy(
            id = null,
            name = "${existingComponent.name}_副本",
            status = 0
        )
        
        val componentId = activityComponentDao.save(newComponent)
        newComponent.id = componentId
        
        val componentDto = convertToDto(newComponent)
        return ApiData.success(componentDto)
    }
    
    /**
     * 转换为DTO
     */
    private fun convertToDto(component: ActivityComponent): ActivityComponentDto {
        val fields = parseFields(component.fields)
        
        return ActivityComponentDto(
            id = component.id,
            name = component.name,
            description = component.description,
            fields = fields,
            status = component.status,
            createTime = TimeUtil.format(component.createTime),
            updateTime = TimeUtil.format(component.updateTime)
        )
    }
    
    /**
     * 转换为实体
     */
    private fun convertToEntity(form: ActivityComponentFormDto): ActivityComponent {
        val fieldsJson = serializeFields(form.fields)
        
        return ActivityComponent(
            name = form.name,
            description = form.description,
            fields = fieldsJson,
            status = form.status
        )
    }
    
    /**
     * 序列化字段/组件配置为JSON字符串
     */
    private fun serializeFields(fields: List<Map<String, Any>>?): String? {
        return try {
            if (fields.isNullOrEmpty()) {
                null
            } else {
                JacksonConfig.objectMapper.writeValueAsString(fields)
            }
        } catch (e: Exception) {
            logger.error("序列化字段/组件配置失败: ${e.message}", e)
            null
        }
    }
    
    /**
     * 解析字段/组件配置JSON字符串
     */
    private fun parseFields(fieldsJson: String?): List<Map<String, Any>>? {
        return try {
            if (fieldsJson.isNullOrBlank()) {
                null
            } else {
                @Suppress("UNCHECKED_CAST")
                val fieldsList = JacksonConfig.objectMapper.readValue(fieldsJson, List::class.java) as? List<Map<String, Any>>
                fieldsList
            }
        } catch (e: Exception) {
            logger.error("解析字段/组件配置失败: ${e.message}", e)
            null
        }
    }
}
